File: src/main/golang/leetcode/editor/cn/[221]Maximal Square.go
package lc

//leetcode submit region begin(Prohibit modification and deletion)
func maximalSquare(matrix [][]byte) (ans int) {
	Max := func(x, y int) int {
		if x > y {
			return x
		}
		return y
	}
	Min := func(x, y int) int {
		if x > y {
			return y
		}
		return x
	}
	m, n := len(matrix), len(matrix[0])
	dp := make([][]int, m+1)
	dp[0] = make([]int, n+1)
	for i := 1; i <= m; i++ {
		dp[i] = make([]int, n+1)
		for j := 1; j <= n; j++ {
			if matrix[i-1][j-1] == '1' {
				dp[i][j] = 1 + Min(dp[i-1][j-1], Min(dp[i-1][j], dp[i][j-1]))
			}
			ans = Max(ans, dp[i][j]*dp[i][j])
		}
	}
	return
}

//leetcode submit region end(Prohibit modification and deletion)

File: src/main/golang/leetcode/editor/cn/[221]Maximal Square.py
from typing import List
# leetcode submit region begin(Prohibit modification and deletion)
class Solution:
    def maximalSquare(self, matrix: List[List[str]]) -> int:
        if not matrix:
            return 0
        n, m = len(matrix), len(matrix[0])
        dp = [[0] * m for _ in range(n)]
        ans = 0
        for i in range(n):
            for j in range(m):
                if matrix[i][j] == '1':
                    dp[i][j] = 1
                    if i > 0 and j > 0:
                        dp[i][j] += min(dp[i - 1][j], dp[i][j - 1], dp[i - 1][j - 1])
                    ans = max(ans, dp[i][j])
        return ans * ans
# leetcode submit region end(Prohibit modification and deletion)
